spark2/src/main/java/org/apache/iceberg/spark/source/Reader.java

```diff
@@ -206,11 +219,17 @@ public StructType readSchema() {
     Broadcast<Table> tableBroadcast = sparkContext.broadcast(SerializableTable.copyOf(table));
 
     List<InputPartition<ColumnarBatch>> readTasks = Lists.newArrayList();
-    for (CombinedScanTask task : tasks()) {
-      readTasks.add(new ReadTask<>(
-          task, tableBroadcast, expectedSchemaString, caseSensitive,
-          localityPreferred, new BatchReaderFactory(batchSize)));
-    }
+    Tasks.foreach(tasks())
+        .stopOnFailure()
+        .executeWith(readTasksInitExecutorService)
+        .run(task -> {
+          InputPartition<ColumnarBatch> readTask = new ReadTask<>(
+              task, tableBroadcast, expectedSchemaString, caseSensitive,
+              localityPreferred, new BatchReaderFactory(batchSize));
+          synchronized (readTasks) {
```

Comment:
Why not use a concurrent list?

Reply:
At first time I wanted to use a concurrent list. But this would change the return value from a normal list to a concurrent list, and the subsequent operation on this list should only be read. Maybe it will affect performance.

Reply:
Or I have fallen into a premature optimization 😂

Reply:
I changed the code to use arrays like in spark3 to avoid synchronization.